controllers/imagerepository_controller.go

```diff
@@ -72,55 +73,85 @@ func (r *ImageRepositoryReconciler) Reconcile(req ctrl.Request) (ctrl.Result, er
 
 	ref, err := name.ParseReference(imageRepo.Spec.Image)
 	if err != nil {
-		imageRepo.Status.LastError = err.Error()
-		if err := r.Status().Update(ctx, &imageRepo); err != nil {
-			return ctrl.Result{}, err
+		status := imagev1alpha1.SetImageRepositoryReadiness(
+			imageRepo,
+			corev1.ConditionFalse,
+			imagev1alpha1.ReconciliationFailedReason,
+			err.Error(),
+		)
+		if err := r.Status().Update(ctx, &status); err != nil {
+			return ctrl.Result{Requeue: true}, err
 		}
 		log.Error(err, "Unable to parse image name", "imageName", imageRepo.Spec.Image)
-		return ctrl.Result{}, nil
+		return ctrl.Result{}, err
```

Comment:
I'm deliberately not returning an error -- the reconciliation _succeeded_, and the result was that the input is considered invalid.

More generally, not reaching a ready state does not mean that reconciliation failed. It may just be that some long-running process is still underway.

Reply:
> I'm deliberately not returning an error -- the reconciliation succeeded, and the result was that the input is considered invalid.

If the spec is invalid we should send an error alert to notification-controller and record it as a failed reconciliation, same way we do for sources https://github.com/fluxcd/source-controller/blob/master/controllers/helmrepository_controller.go#L176

Are you suggesting to change the way validation errors are recorded and ignore them?

> More generally, not reaching a ready state does not mean that reconciliation failed. 

Ok but in this case it's a validation error. We could have a dedicated status reason like https://github.com/fluxcd/source-controller/blob/master/api/v1alpha1/condition_types.go#L62

> It may just be that some long-running process is still underway.

We signal this by setting the ready condition to unknown and the reason to progressing. In that case we would not return an error.

Reply:
Yes that's a fair point, invalid input is different to just failing to make progress. This is what validation hooks are for I presume, though a bit of a sledgehammer.

It's the owner of the resource that needs to see the problem, rather than the operator of the controller, so it feels like a condition is needed either way.

If the image string parsed OK, but wasn't accessible (either failure to authenticate, or didn't exist -- which may not be distinguishable cases), I'd use a condition to signal that the resource will not make progress without intervention. Arguably, a misshapen image string is closer to that situation than to a reconciliation failure. But it's a judgement call, I think.

Reply:
I've added an URL invalid reason as we do in source controller.

Reply:
My conclusion is: let's have a dedicated status reason for being unable to scan the repo. There's no point in requeuing it, because it needs to be changed before it'll work; and, it's not an unexpected failure of the controller, and it doesn't need to be logged; so, no reason to return an error from Reconcile.

Reply:
I've changed it to return no error.

Reply:
(changed back, subsequently; see https://github.com/fluxcd/image-reflector-controller/pull/24#issuecomment-677647182)